### DISCUSSION TITLE
DNS Resolver - Zephyr handling non-zero ‘Z’ bits in response header

### DISCUSSION THREAD

**Discussion by Gatchers** [zephyrproject-rtos/zephyr#52480] [RFC]
Hi!

Mid-term user, new poster, sorry!

Zephyr faults DNS responses with a non-zero ‘Z’ field. Servers such as the Dual DNS/DHCP server currently serve undefined bits in this field, but I guess one day these bits may be defined and used?

I guess I’m wondering what the strategy is here. At the moment, this explains why Zephyr clients can (not always as the bits are undefined!) fail here.

Thanks,

Gareth

**Comment by GeorgeCGV** [comment]
Maybe Zephyr implementation strictly follows [RFC1035](https://datatracker.ietf.org/doc/html/rfc1035):
```
Z               Reserved for future use.  Must be zero in all queries
                and responses.
```
that clearly states the requirement of bits to be `0`.

Later on [RFC2535 ](https://datatracker.ietf.org/doc/html/rfc2535#section-6.1) uses [RFC1035](https://datatracker.ietf.org/doc/html/rfc1035) `Z` field for security purposes (_splits it into AD, CD, and Z bits_). 

It doesn't look like the implementation considers AD/CD bits described in [RFC2535 ](https://datatracker.ietf.org/doc/html/rfc2535#section-6.1):

https://github.com/zephyrproject-rtos/zephyr/blob/c8942afe6d63c0d8f78898c421e063a46fb66a4e/subsys/net/lib/dns/dns_pack.h#L234-L237

The `dns_unpack_response_header` will result in `-EINVAL` if `Z` bits aren't `0`:

https://github.com/zephyrproject-rtos/zephyr/blob/ac7300611c0e19748d0c48b8ca2dfdd03578797a/subsys/net/lib/dns/dns_pack.c#L204-L206

That leads to `DNS_EAI_FAIL` within `dns_validate_msg`:

https://github.com/zephyrproject-rtos/zephyr/blob/c8942afe6d63c0d8f78898c421e063a46fb66a4e/subsys/net/lib/dns/resolve.c#L512-L516

**Comment by Gatchers** [comment]
Thank you - yes, this corresponds with what I found via debugging. I am interested in what the plan is here.

**Reply by Gatchers** [reply]
Also well done! You traced back exactly what I had to find via debug, thank you!

**Reply by GeorgeCGV** [reply]
I am not sure if there is a plan beyond [RFC1035](https://datatracker.ietf.org/doc/html/rfc1035) support.
Contributions are always welcomed and appreciated.
